### ISSUE TITLE
Running a package with embedded jx

### ISSUE THREAD

**Issue by zamb3zi** [jxcore/jxcore#283]
I'm developing a windows application where I would like to embed jxcore. I've built jx.dll and I've had a look through jx.h/jx.cc.
My question is: can I compile a .jx package and execute it via the public interface? I would like to embed the package as a resource in my application which will be copy protected.

**Comment by zamb3zi** [comment]
From what I can tell, something like the following may need to be added to the public interface to allow this:
```c++
JXCORE_EXTERN(void) 
JX_DefineMainJXFile(const void *data, size_t size);

JXCORE_EXTERN(void) 
JX_DefineJXFile(const char *name, const void *data, size_t size);
```
Or am I on the wrong track? Thanks.

**Comment by obastemur** [comment]
Just to be on the same track; 
```
1 - package a node app into a jx file
2 - develop a separate solution that embeds jxcore
3 - use that jx file with this solution
```

Please let me know if I understood it wrong.

**Comment by zamb3zi** [comment]
Yes, that's it exactly. The reason is that I want jx-file protection, but I can't delegate the main executable to jxcore. I still need to embed it. So I would like to deliver the jx-file in-memory from a protected [resource](https://msdn.microsoft.com/en-us/library/windows/desktop/ms648046%28v=vs.85%29.aspx).

**Comment by obastemur** [comment]
Why don't you require that JX file. 

```
JX_DefineMainFile("require('myfile.jx');");
```

**Comment by zamb3zi** [comment]
Sure, but then I have to install `myfile.jx` on disk and it needs to be `"library": true`. It wouldn't fall under the copy protection of my main application as it would if it were a compiled resource delivered in-memory to jxcore. One option for me would be to use the [RequireJS optimizer](http://requirejs.org/docs/optimization.html) and then embed these as text-file resources delivered to `JX_DefineMainFile`, but it seems a shame not to use the jx packager as it would be perfect if only I could deliver the jx data in memory.

**Comment by obastemur** [comment]
I see your point and it makes sense but I'm not sure the priority of this item. If you can contribute, It could be great! Nonetheless, I can turn back to this feature after completing the missing interface items from jx-ni.

**Comment by zamb3zi** [comment]
No problem. Yes, it could be a good project for me.

**Comment by kmgkv** [comment]
Hi obastemur,
I have a similar use case, I'm embedding JXCore in iOS and Android application and would like to keep my scripts content encrypted in it, so I created the JX package of script that I'm bundling in iOS and Android applications.

I would like to know in details what kind of encryption scheme is used for packaging? How secure it is?

So far i have tested <b>JX_DefineMainFile("require('myfile.jx');");</b> on iOS, it works, I'm about to try this on Android...

**Comment by kmgkv** [comment]
Also, given the fact that on Android, we need some kind of redirection in main.js to support require statement (as explained in your android sample), so i was wondering how <b>JX_DefineMainFile("require('myfile.jx');");</b> would work on Andorid

**Comment by obastemur** [comment]
`so i was wondering how JX_DefineMainFile("require('myfile.jx');"); would work on Andorid`

It should work as it is. 

`I would like to know in details what kind of encryption scheme is used for packaging? How secure it is?`

We stopped using encryption for packages. One of the items on my hand for mobile apps is to compile source codes into packages. 

`I have a similar use case, I'm embedding JXCore in iOS and Android application and would like to keep my scripts content encrypted in it, so I created the JX package of script that I'm bundling in iOS and Android applications.`

Seems like @zamb3zi is going to help with the implementation. On the other hand, looks like you have your solution for now.

**Comment by kmgkv** [comment]
So can we say that, a packaged node app is nothing but the compiled bundle (compiling JS files into native code) of all the JS files into single .JX file.. Content of .JX package file is not encrypted????

**Comment by ktrzeciaknubisa** [comment]
@kmgkv Yes, not encrypted. This is since JXcore is open sourced, because the previous encryption approach could not be guaranteed any more once everyone can look into the sources and track the encryption/decryption methods.

**Comment by kmgkv** [comment]
@ktrzeciaknubisa thanks for info, just quickly looked into node.js source file and jxutils_wraps.cc, seems like all we are doing here is compressing the files and packaging all of them into another compressed JSON which is the final .JX file!!!!

**Comment by obastemur** [comment]
It's a bit more than compressing (we do not use zlib from the deps for it) yet securing the content part will be handled separately. This is one of the things hopefully we will be taking care of soon. 

On the encryption side, I think slightly different. Making JXcore open source doesn't make much difference on giving encryption support. It will be a pre-compile time parameter along with the --compress-internals key.

for example on unix; (same applies to vcbuilt.bat on windows)
```
./configure --compress-internals --encrypt-packages=<key file here>
sudo make
```

As a result, your package will be working only with the jx binary that you have compiled. Wihtout hacking into jx binary you've compiled, it won't be easy to see the contents of your package.

The only problem here is that I don't know how much it will be useful for general use cases. any comment on this might help me to design it

**Comment by kmgkv** [comment]
@obastemur I liked your approach however, to make it useful for general public, I guess it will be good if we can think of following design:
<ul>
  <li>Add another sub-option for JX package command to specify public part of the asym key. For example : JX package main_file_name pack_name -key pub_key_path</li>
<li>JX executable will continue to package the node package like the way it is doing as of today, but at the end, it will use the public key to encrypt entire compressed blob</li>
<li>Whether or not the package is encrypted, that info can be maintained in .JXP project file,so taht while using the .JX package we can accordingly decrypt or not</li>
  <li>Expose one more JXCore API to specify the private key before initializing the engine.</li>
<li>Make use of specified priv key to decrypt the package</li>
</ul>
So, by doing this you are providing option to the JXCore embedders that they can encrypt and decrypt the script using their own pub-priv key pair.. Embedders should make sure to keep the keys safe.. Also, this provides a way to run multiple engine instances, each running different encryption scheme!

**Comment by obastemur** [comment]
@kmgkv sounds good and this approach may solve the problem for embedded apps. Perhaps same private - public key thing can also work for non-embedded jx packages too. Compile jx binary using your private key and deliver the packages with public key embedded inside.

As a result those packages would only work on the right jx binary without compromising the private key. 
I've some edge cases though;

- There is a need for binary shuffling in order to prevent (or make it much harder) for an attacker to estimate your private key. Keep the input files on a jx package intact, each compile operation should generate randomly ordered jx package binary data. 

- compression should take a place before encryption. (This will be increasing the package size a bit comparing to now)

- all this stuff must require --compress-internals configuration switch 

what else..

**Comment by kmgkv** [comment]
@obastemur agreed and makes sense. Binary shuffling will reduce the possibility of key encryption...we should keep this discussion running and finally when we have it all we can implement it!

**Comment by kmgkv** [comment]
@obastemur I have completed the implementation of for running encrypted/integrity protected jxcore packages for embedded application.. 

Is this is something you would like us to contribute it to main repo, let me know what your thoughts are, I can explain you the design and approach.. It is pretty much same like I discussed above.. however I'm happy to explain the full details..

Let me know your thought...
Thanks,
kmgkv

**Comment by obastemur** [comment]
@kmgkv feel free to submit a PR and lets discuss there. Thanks.
